### PR TITLE
Add `--config` and `--source` flags to push

### LINKED ISSUE #1489: odo push --source, odo push --config
from https://github.com/redhat-developer/odo/issues/1488

But the above will take time to implement, for beta release we could just do the following:

-  odo push --source will just copy files and trigger build
- odo push --config will just apply config changes
- by default, odo push will do both (later we will make this smarter as described above)

### PR DESCRIPTION
**Pull request by anmolbabu**
## What is the purpose of this change? What does it change?

This commit adds:
1. `--config` flag to update the component with component settings
2. `--source` flag to push only the source to an already existing
   component.
3. If neither `--config` nor `--source` flags are passed, update
   both source and configuration onto the component

fixes #1489
Signed-off-by: anmolbabu <anmolbudugutta@gmail.com>

## How to test changes?

* `odo push`
* `odo push --source`
* `odo push --config`

The above 3 with `--context` also should behave with changes above

### CODE REVIEW

**Comment by anmolbabu** [comment]
ToDo:

- [ ] Add e2e tests to cover `--config` and `--source`

**Comment by girishramnani** [comment]
assuming tests will be added in the new structure?

File: pkg/odo/cli/component/push.go [hunk 1]
```diff
@@ -101,22 +101,40 @@ func (po *PushOptions) Complete(name string, cmd *cobra.Command, args []string)
...
+	}
+
+	if !isCmpExists && po.pushSource && !po.pushConfig {
+		return fmt.Errorf("Component %s does not exist and hence cannot push only source. Please use `odo push` without any flags or with both `--source` and `--config` flags", po.localConfig.GetName())
+	}
+
+	return nil
 }
 
 func (po *PushOptions) createCmpIfNotExistsAndApplyCmpConfig(stdout io.Writer) error {
```
Comment (girishramnani) [inline pkg/odo/cli/component/push.go]: this might be a good time to break this function into two` createCmpIfNotExists` and `applyCmdConfig`. there is alot of responsibility on a single function

**Comment by surajnarwade** [comment]
Tried, works for me:)

File: pkg/odo/cli/component/push.go (hunk 1 above)
Comment (anmolbabu) [reply pkg/odo/cli/component/push.go]: But applyconfig is just a single function call the implementation is already out in a different function

File: pkg/odo/cli/component/push.go [hunk 2]
```diff
@@ -245,6 +278,8 @@ func NewCmdPush(name, fullName string) *cobra.Command {
 	pushCmd.Flags().StringVarP(&po.componentContext, "context", "c", "", "Use given context directory as a source for component settings")
```
Comment (surajnarwade) [inline pkg/odo/cli/component/push.go]: small suggestion, can you add autocompletion for `--context` flag, it can be different PR also.

```
completion.RegisterCommandFlagHandler(pushCmd, "context", completion.FileCompletionHandler)
```

Comment (anmolbabu) [reply pkg/odo/cli/component/push.go]: sure

File: tests/e2e/helper/helper_oc.go [hunk 3]
```diff
@@ -36,3 +36,21 @@ func OcGetCurrentProject() string {
 	}
 	return ""
 }
+
+// CheckCmdOpInRemoteCmpPod runs the provided command on remote component pod and returns the return value of command output handler function passed to it
+func CheckCmdOpInRemoteCmpPod(cmpName string, appName string, cmd string, checkOp func(cmdOp string, err error) bool) bool {
+	cmpDCName := fmt.Sprintf("%s-%s", cmpName, appName)
+	podName := CmdShouldPass(fmt.Sprintf("oc get pods | grep %s | awk '{print $1}' | tr -d '\n'", cmpName))
```
Comment (kadel) [inline tests/e2e/helper/helper_oc.go]: we can't use this. It uses awk, and grep and those won't work on Windows

Comment (kadel) [reply tests/e2e/helper/helper_oc.go]: you should be able to use something like this `oc get pods --selector "deploymentconfig=%s" jsonpath='{.metadata.name}'`

Comment (anmolbabu) [reply tests/e2e/helper/helper_oc.go]: Ack

File: Makefile [hunk 4]
```diff
@@ -111,9 +111,9 @@ endif
 .PHONY: test-cmp-e2e
 test-cmp-e2e:
 ifdef TIMEOUT
-	go test -v github.com/openshift/odo/tests/e2e --ginkgo.focus="odoCmpE2e" -ginkgo.succinct -timeout $(TIMEOUT)
+	go test -v github.com/openshift/odo/tests/e2e --ginkgo.focus="odoCmpE2e" -ginkgo.v -timeout $(TIMEOUT)
```
Comment (kadel) [inline Makefile]: do we still need a verbose flag?
For failed tests, you will see verbose output even with `succinct`

Comment (anmolbabu) [reply Makefile]: Yeah I am having it only until the tests are green I'll revert the last commit that contains this completely before merge

**Comment by kadel** [comment]
I keep getting this after dong `odo create`, `odo push`, `odo push --source`

```
odo push --config
 ✓  Checking component
 ✓  Checking component version
 ✓  Applying component settings to component: mycomp
 ✗  Failed to update config to component deployed. Error Timed out waiting for annotation to update
github.com/openshift/odo/pkg/occlient.(*Client).WaitAndGetDC
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/occlient/occlient.go:1672
github.com/openshift/odo/pkg/occlient.(*Client).PatchCurrentDC
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/occlient/occlient.go:1309
github.com/openshift/odo/pkg/occlient.(*Client).UpdateDCToSupervisor
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/occlient/occlient.go:1500
github.com/openshift/odo/pkg/component.Update
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/component/component.go:1010
github.com/openshift/odo/pkg/component.ApplyConfig
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/component/component.go:531
github.com/openshift/odo/pkg/odo/cli/component.(*PushOptions).createCmpIfNotExistsAndApplyCmpConfig
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/odo/cli/component/push.go:166
github.com/openshift/odo/pkg/odo/cli/component.(*PushOptions).Run
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/odo/cli/component/push.go:183
github.com/openshift/odo/pkg/odo/genericclioptions.GenericRun
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/odo/genericclioptions/runnable.go:17
github.com/openshift/odo/pkg/odo/cli/component.NewCmdPush.func1
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/odo/cli/component/push.go:268
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).execute
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:702
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).ExecuteC
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:783
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).Execute
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:736
main.main
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/cmd/odo/odo.go:73
runtime.main
	/usr/local/Cellar/go/1.12.1/libexec/src/runtime/proc.go:200
runtime.goexit
	/usr/local/Cellar/go/1.12.1/libexec/src/runtime/asm_amd64.s:1337
unable to wait for DeploymentConfig mycomp-myapp to update
github.com/openshift/odo/pkg/occlient.(*Client).PatchCurrentDC
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/occlient/occlient.go:1311
github.com/openshift/odo/pkg/occlient.(*Client).UpdateDCToSupervisor
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/occlient/occlient.go:1500
github.com/openshift/odo/pkg/component.Update
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/component/component.go:1010
github.com/openshift/odo/pkg/component.ApplyConfig
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/component/component.go:531
github.com/openshift/odo/pkg/odo/cli/component.(*PushOptions).createCmpIfNotExistsAndApplyCmpConfig
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/odo/cli/component/push.go:166
github.com/openshift/odo/pkg/odo/cli/component.(*PushOptions).Run
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/odo/cli/component/push.go:183
github.com/openshift/odo/pkg/odo/genericclioptions.GenericRun
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/odo/genericclioptions/runnable.go:17
github.com/openshift/odo/pkg/odo/cli/component.NewCmdPush.func1
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/odo/cli/component/push.go:268
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).execute
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:702
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).ExecuteC
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:783
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).Execute
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:736
main.main
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/cmd/odo/odo.go:73
runtime.main
	/usr/local/Cellar/go/1.12.1/libexec/src/runtime/proc.go:200
runtime.goexit
	/usr/local/Cellar/go/1.12.1/libexec/src/runtime/asm_amd64.s:1337
unable to update the current DeploymentConfig mycomp-myapp
github.com/openshift/odo/pkg/occlient.(*Client).UpdateDCToSupervisor
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/occlient/occlient.go:1509
github.com/openshift/odo/pkg/component.Update
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/component/component.go:1010
github.com/openshift/odo/pkg/component.ApplyConfig
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/component/component.go:531
github.com/openshift/odo/pkg/odo/cli/component.(*PushOptions).createCmpIfNotExistsAndApplyCmpConfig
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/odo/cli/component/push.go:166
github.com/openshift/odo/pkg/odo/cli/component.(*PushOptions).Run
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/odo/cli/component/push.go:183
github.com/openshift/odo/pkg/odo/genericclioptions.GenericRun
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/odo/genericclioptions/runnable.go:17
github.com/openshift/odo/pkg/odo/cli/component.NewCmdPush.func1
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/odo/cli/component/push.go:268
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).execute
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:702
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).ExecuteC
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:783
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).Execute
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:736
main.main
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/cmd/odo/odo.go:73
runtime.main
	/usr/local/Cellar/go/1.12.1/libexec/src/runtime/proc.go:200
runtime.goexit
	/usr/local/Cellar/go/1.12.1/libexec/src/runtime/asm_amd64.s:1337
unable to update DeploymentConfig for mycomp component
github.com/openshift/odo/pkg/component.Update
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/component/component.go:1016
github.com/openshift/odo/pkg/component.ApplyConfig
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/component/component.go:531
github.com/openshift/odo/pkg/odo/cli/component.(*PushOptions).createCmpIfNotExistsAndApplyCmpConfig
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/odo/cli/component/push.go:166
github.com/openshift/odo/pkg/odo/cli/component.(*PushOptions).Run
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/odo/cli/component/push.go:183
github.com/openshift/odo/pkg/odo/genericclioptions.GenericRun
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/odo/genericclioptions/runnable.go:17
github.com/openshift/odo/pkg/odo/cli/component.NewCmdPush.func1
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/pkg/odo/cli/component/push.go:268
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).execute
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:702
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).ExecuteC
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:783
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).Execute
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:736
main.main
	/Users/tkral/Code/goenv/odo/src/github.com/openshift/odo/cmd/odo/odo.go:73
runtime.main
	/usr/local/Cellar/go/1.12.1/libexec/src/runtime/proc.go:200
runtime.goexit
	/usr/local/Cellar/go/1.12.1/libexec/src/runtime/asm_amd64.s:1337

tmp/odo-examples/fuse-rest-http-booster  07d1c28 ✗
```

File: tests/e2e/component.go [hunk 5]
```diff
@@ -470,4 +472,228 @@ func componentTests(componentCmdPrefix string) {
...
+				var _ = JustAfterEach(func() {
+					helper.OcSwitchProject(originalProject)
+				})
+
+				It("create local nodejs component and push source and code separately", func() {
+					appName := "nodejs-push-test"
+					cmpName := "nodejs"
+					helper.CopyExample(filepath.Join("source", "nodejs"), context)
+
+					helper.CmdShouldPass("odo component create nodejs " + cmpName + " --app " + appName)
```
Comment (kadel) [inline tests/e2e/component.go]: you have this inside `Context("when --context is not used",` but you are not using context!

File: tests/e2e/component.go [hunk 6]
```diff
@@ -470,4 +472,228 @@ func componentTests(componentCmdPrefix string) {
...
+		// Clean up after the test
+		// This is run after every Spec (It)
+		var _ = AfterEach(func() {
+			helper.OcDeleteProject(project)
+			helper.DeleteDir(context)
+		})
+
+		// we will be testing components that are created from the current directory
+		// switch to the clean context dir before each test
+		var _ = JustBeforeEach(func() {
```
Comment (kadel) [inline tests/e2e/component.go]: this shouldn't be here

File: tests/e2e/component.go [hunk 7]
```diff
@@ -470,4 +472,228 @@ func componentTests(componentCmdPrefix string) {
 			waitForDeleteCmd("odo project list", projName)
 		})
 	})
+
+	var _ = Context("when component is in the current directory", func() {
```
Comment (kadel) [inline tests/e2e/component.go]: this should have some meaningful name, like `Testing odo push with --source and --config flags`

File: tests/e2e/component.go (hunk 6 above)
Comment (kadel) [reply tests/e2e/component.go]: it should be inside  Context("when --context is not used", func()

File: tests/e2e/component.go [hunk 8]
```diff
@@ -474,4 +476,173 @@ func componentTests(componentCmdPrefix string) {
...
+		})
+
+		// Clean up after the test
+		// This is run after every Spec (It)
+		AfterEach(func() {
+			helper.OcDeleteProject(project)
+			helper.DeleteDir(context)
+		})
+
+		Context("when project from KUBECONFIG is used", func() {
```
Comment (kadel) [inline tests/e2e/component.go]: this should be renamed now to something like "when --project is used"

File: tests/e2e/component.go [hunk 9]
```diff
@@ -474,4 +476,173 @@ func componentTests(componentCmdPrefix string) {
...
+			// we will switching project to new one for each test
+			JustBeforeEach(func() {
```
Comment (kadel) [inline tests/e2e/component.go]: this should be removed now

File: tests/e2e/component.go [hunk 10]
```diff
@@ -474,4 +476,173 @@ func componentTests(componentCmdPrefix string) {
...
+				originalProject = helper.OcGetCurrentProject()
+				helper.OcSwitchProject(project)
+			})
+
+			// go back to original project after each test
+			JustAfterEach(func() {
```
Comment (kadel) [inline tests/e2e/component.go]: this should be removed now

File: tests/e2e/component.go (hunk 8 above)
Comment (anmolbabu) [reply tests/e2e/component.go]: ack

File: tests/e2e/component.go (hunk 9 above)
Comment (anmolbabu) [reply tests/e2e/component.go]: ack

File: tests/e2e/component.go (hunk 10 above)
Comment (anmolbabu) [reply tests/e2e/component.go]: ack

File: tests/e2e/helper/helper_run.go [hunk 11]
```diff
@@ -26,7 +26,7 @@ func cmdRunner(cmdS string) (string, string, int) {
 	stderr := string(session.Err.Contents())
 	exitCode := session.ExitCode()
 
-	fmt.Fprintf(GinkgoWriter, "Result: \n stdout: %s\n stderr:%s \n exitcode: %d \n", stdout, stderr, exitCode)
+	//fmt.Fprintf(GinkgoWriter, "Result: \n stdout: %s\n stderr:%s \n exitcode: %d \n", stdout, stderr, exitCode)
```
Comment (surajnarwade) [inline tests/e2e/helper/helper_run.go]: remove the command if not needed

File: tests/e2e/helper/helper_run.go [hunk 12]
```diff
@@ -38,3 +38,11 @@ func CmdShouldPass(cmd string) string {
 	Expect(exitcode).To(Equal(0))
 	return strings.TrimSpace(stdout)
 }
+
+// CmdShouldFail command needs to retrun non 0 as en exit code
```
Comment (surajnarwade) [inline tests/e2e/helper/helper_run.go]: small typo, `s/retrun/return`

**Comment by surajnarwade** [comment]
@kadel ,@anmolbabu, 

After `odo create nodejs` and `odo push`, I did following,

```
$ odo push --config 
 ✓  Checking component
 ✓  Checking component version
 ✓  Applying component settings to component: nodejs-umic
 ✓  The component nodejs-umic was updated successfully
 ✓  Successfully created component with name: nodejs-umic
```

last line says, `Successfully created component with name: nodejs-umic` whereas we just updated

File: tests/e2e/helper/helper_run.go (hunk 11 above)
Comment (anmolbabu) [reply tests/e2e/helper/helper_run.go]: I'll have it for debugging

File: tests/e2e/helper/helper_run.go (hunk 12 above)
Comment (anmolbabu) [reply tests/e2e/helper/helper_run.go]: :)

**Comment by anmolbabu** [comment]
> @kadel ,@anmolbabu,
> 
> After `odo create nodejs` and `odo push`, I did following,
> 
> ```
> $ odo push --config 
>  ✓  Checking component
>  ✓  Checking component version
>  ✓  Applying component settings to component: nodejs-umic
>  ✓  The component nodejs-umic was updated successfully
>  ✓  Successfully created component with name: nodejs-umic
> ```
> last line says, `Successfully created component with name: nodejs-umic` whereas we just updated

In this case it is actually created itself

**Comment by surajnarwade** [comment]
@anmolbabu , I tried with latest changes and it is working for me now :tada: 

```
$ odo push
 ✓  Checking component
 ✓  Checking component version
 ✓  Applying component settings to component: nodejs-naql
 ✓  The component nodejs-naql was updated successfully
 ✓  Successfully updated component with name: nodejs-naql
 ✓  Pushing changes to component: nodejs-naql of type local
 ✓  Waiting for component to start
 ✓  Copying files to component
 ✓  Building component
 ✓  Changes successfully pushed to component: nodejs-naql
```

```
$ odo push --source --config
 ✓  Checking component
 ✓  Checking component version
 ✓  Applying component settings to component: nodejs-naql
 ✓  The component nodejs-naql was updated successfully
 ✓  Successfully updated component with name: nodejs-naql
 ✓  Pushing changes to component: nodejs-naql of type local
 ✓  Waiting for component to start
 ✓  Copying files to component
 ✓  Building component
 ✓  Changes successfully pushed to component: nodejs-naql
```

```
$ odo push --config 
 ✓  Checking component
 ✓  Checking component version
 ✓  Applying component settings to component: nodejs-naql
 ✓  The component nodejs-naql was updated successfully
 ✓  Successfully updated component with name: nodejs-naql
```

```
$ odo push --source
 ✓  Checking component
 ✓  Checking component version
 ✓  Pushing changes to component: nodejs-naql of type local
 ✓  Waiting for component to start
 ✓  Copying files to component
 ✓  Building component
 ✓  Changes successfully pushed to component: nodejs-naql
```

**Comment by kadel** [comment]
/test e2e-core
--

File: tests/e2e/component.go [hunk 13]
```diff
@@ -474,4 +476,166 @@ func componentTests(componentCmdPrefix string) {
...
+				JustAfterEach(func() {
+					helper.Chdir(originalDir)
+				})
+
+				It("create local nodejs component and push source and code separately", func() {
+					appName := "nodejs-push-test"
+					cmpName := "nodejs"
+					helper.CopyExample(filepath.Join("source", "nodejs"), context)
+
+					helper.CmdShouldPass("odo component create nodejs " + cmpName + " --app " + appName + " --project " + project)
```
Comment (amitkrout) [inline tests/e2e/component.go]: Please replace the occurrence of ```odo component``` with ```componentCmdPrefix``` just to make sure component and component sub tests are verified in the new template.

File: tests/e2e/component.go [hunk 14]
```diff
@@ -474,4 +476,166 @@ func componentTests(componentCmdPrefix string) {
...
+
+			Context("when --context is used", func() {
+				// don't need to switch to any dir here, as this test should use --context flag
+
+				It("create local nodejs component and push source and code separately", func() {
+					appName := "nodejs-push-context-test"
+					cmpName := "nodejs"
+					helper.CopyExample(filepath.Join("source", "nodejs"), context)
+
+					helper.CmdShouldPass("odo component create nodejs " + cmpName + " --context " + context + " --app " + appName + " --project " + project)
```
Comment (amitkrout) [inline tests/e2e/component.go]: Here too

File: tests/e2e/component.go [hunk 15]
```diff
@@ -474,4 +476,166 @@ func componentTests(componentCmdPrefix string) {
...
+					)
+					Expect(remoteCmdExecPass).To(Equal(true))
+				})
+
+				It("create local nodejs component and push source and code at once", func() {
+					appName := "nodejs-push-context-test"
+					cmpName := "nodejs-push-atonce"
+					helper.CopyExample(filepath.Join("source", "nodejs"), context)
+
+					helper.CmdShouldPass("odo component create nodejs " + cmpName + " --app " + appName + " --context " + context + " --project " + project)
```
Comment (amitkrout) [inline tests/e2e/component.go]: Here also

File: tests/e2e/component.go [hunk 16]
```diff
@@ -474,4 +476,166 @@ func componentTests(componentCmdPrefix string) {
...
+					)
+					Expect(remoteCmdExecPass).To(Equal(true))
+				})
+
+				It("create local nodejs component and push source and code at once", func() {
+					appName := "nodejs-push-test"
+					cmpName := "nodejs-push-atonce"
+					helper.CopyExample(filepath.Join("source", "nodejs"), context)
+
+					helper.CmdShouldPass("odo component create nodejs " + cmpName + " --app " + appName + " --project " + project)
```
Comment (amitkrout) [inline tests/e2e/component.go]: Here too
